File: Form1.vb
Imports System.ComponentModel.DataAnnotations
Imports MySql.Data.MySqlClient

Public Class Form1
    ' Inisialisasi koneksi ke database MySQL
    Dim conn As New MySqlConnection("server=localhost;port=3306;username=root;password=;database=curd")
    Dim dr As MySqlDataReader

    ' Saat form dimuat, data langsung ditampilkan di DataGridView
    Private Sub Form1_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        DGV_load()
    End Sub

    ' Tombol Simpan untuk menyimpan data ke tabel tbl_curd
    Private Sub btn_save_Click(sender As Object, e As EventArgs) Handles btn_save.Click
        Dim cmd As New MySqlCommand("INSERT INTO `tbl_curd`(`NOPRODUK`,`NAMAPRODUK`, `HARGA`, `GRUP`, `EXPDATE`, `STATUS`) VALUES (@NOPRODUK,@NAMAPRODUK,@HARGA,@GRUP,@EXPDATE,@STATUS)", conn)
        ' Mengisi parameter dengan data dari form
        cmd.Parameters.AddWithValue("@NOPRODUK", txt_prono.Text)
        cmd.Parameters.AddWithValue("@NAMAPRODUK", txt_Proname.Text)
        cmd.Parameters.AddWithValue("@HARGA", CDec(txt_price.Text))
        cmd.Parameters.AddWithValue("@GRUP", combo_proGroup.Text)
        cmd.Parameters.AddWithValue("@EXPDATE", CDate(exp_datepicker.Value))
        cmd.Parameters.AddWithValue("@STATUS", status_checkbox.Checked)

        Try
            ' Menjalankan perintah INSERT
            conn.Open()
            cmd.ExecuteNonQuery()
            MessageBox.Show("Record Save Success", "CURD", MessageBoxButtons.OK, MessageBoxIcon.Information)
        Catch ex As Exception
            ' Menangani error saat menyimpan
            MessageBox.Show("Record Save Failed: " & ex.Message, "ERROR", MessageBoxButtons.OK, MessageBoxIcon.Asterisk)
        Finally
            conn.Close()
        End Try

        ' Setelah simpan, bersihkan form dan refresh data
        clear()
        DGV_load()
    End Sub

    ' Prosedur untuk memuat data dari database ke DataGridView
    Private Sub DGV_load()
        DataGridView1.Rows.Clear()
        Try
            conn.Open()
            Dim cmd As New MySqlCommand("SELECT * FROM `tbl_curd`", conn)
            dr = cmd.ExecuteReader
            While dr.Read
                ' Tambahkan setiap baris dari hasil query ke DataGridView
                DataGridView1.Rows.Add(dr.Item("NOPRODUK"), dr.Item("NAMAPRODUK"), dr.Item("HARGA"), dr.Item("GRUP"), dr.Item("EXPDATE"), Format(CBool(dr.Item("STATUS"))))
            End While
            dr.Close()
        Catch ex As Exception
            MsgBox("Error loading data: " & ex.Message)
        Finally
            conn.Close()
        End Try
    End Sub

    ' Tombol untuk menghapus isian form
    Private Sub btn_clear_Click(sender As Object, e As EventArgs) Handles btn_clear.Click
        clear()
    End Sub

    Private Sub clear()
        txt_prono.Clear()
        txt_Proname.Clear()
        txt_price.Clear()
        combo_proGroup.Text = ""
        exp_datepicker.Value = DateTime.Now
        status_checkbox.CheckState = CheckState.Unchecked

        ' Aktifkan kembali textbox dan tombol simpan
        txt_prono.ReadOnly = False
        btn_save.Enabled = True
    End Sub

    ' Saat pengguna mengklik data di DataGridView, tampilkan di input form
    Private Sub DataGridView1_CellClick(sender As Object, e As DataGridViewCellEventArgs) Handles DataGridView1.CellClick
        txt_prono.Text = DataGridView1.CurrentRow.Cells(0).Value
        txt_Proname.Text = DataGridView1.CurrentRow.Cells(1).Value
        txt_price.Text = DataGridView1.CurrentRow.Cells(2).Value
        combo_proGroup.Text = DataGridView1.CurrentRow.Cells(3).Value
        exp_datepicker.Text = DataGridView1.CurrentRow.Cells(4).Value
        status_checkbox.Checked = DataGridView1.CurrentRow.Cells(5).Value

        ' Buat NOPRODUK tidak bisa diubah, dan matikan tombol simpan
        txt_prono.ReadOnly = True
        btn_save.Enabled = False
    End Sub

    ' Prosedur untuk mengedit data yang dipilih
    Sub Edit()
        Dim cmd As New MySqlCommand("UPDATE `tbl_curd` SET `NAMAPRODUK` = @NAMAPRODUK, `HARGA` = @HARGA, `GRUP` = @GRUP, `EXPDATE` = @EXPDATE, `STATUS` = @STATUS WHERE `NOPRODUK` = @NOPRODUK", conn)
        cmd.Parameters.Clear()
        cmd.Parameters.AddWithValue("@NOPRODUK", txt_prono.Text)
        cmd.Parameters.AddWithValue("@NAMAPRODUK", txt_Proname.Text)
        cmd.Parameters.AddWithValue("@HARGA", CDec(txt_price.Text))
        cmd.Parameters.AddWithValue("@GRUP", combo_proGroup.Text)
        cmd.Parameters.AddWithValue("@EXPDATE", CDate(exp_datepicker.Value))
        cmd.Parameters.AddWithValue("@STATUS", status_checkbox.Checked)

        Try
            conn.Open()
            cmd.ExecuteNonQuery()
            MessageBox.Show("Record Update Success", "CURD", MessageBoxButtons.OK, MessageBoxIcon.Information)
        Catch ex As Exception
            MessageBox.Show("Record Update Failed: " & ex.Message, "ERROR", MessageBoxButtons.OK, MessageBoxIcon.Asterisk)
        Finally
            conn.Close()
        End Try

        clear()
        DGV_load()
    End Sub

    ' Tombol untuk memanggil prosedur Edit
    Private Sub btn_edit_Click(sender As Object, e As EventArgs) Handles btn_edit.Click
        Edit()
    End Sub

    ' Prosedur untuk menghapus data
    Private Sub Delete()
        If MsgBox("Are you sure you want to delete this record?", MsgBoxStyle.YesNo, "Delete") = MsgBoxResult.Yes Then
            Dim cmd As New MySqlCommand("DELETE FROM `tbl_curd` WHERE `NOPRODUK` = @NOPRODUK", conn)
            cmd.Parameters.Clear()
            cmd.Parameters.AddWithValue("@NOPRODUK", txt_prono.Text)
            Try
                conn.Open()
                cmd.ExecuteNonQuery()
                MessageBox.Show("Record Delete Success", "CURD", MessageBoxButtons.OK, MessageBoxIcon.Information)
            Catch ex As Exception
                MessageBox.Show("Record Delete Failed: " & ex.Message, "ERROR", MessageBoxButtons.OK, MessageBoxIcon.Asterisk)
            Finally
                conn.Close()
            End Try
            clear()
            DGV_load()
        End If
    End Sub

    ' Tombol hapus memanggil prosedur Delete
    Private Sub btn_delete_Click(sender As Object, e As EventArgs) Handles btn_delete.Click
        Delete()
    End Sub

    ' Prosedur pencarian data berdasarkan NOPRODUK atau NAMAPRODUK
    Private Sub txt_search_TextChanged(sender As Object, e As EventArgs) Handles txt_search.TextChanged
        DataGridView1.Rows.Clear()
        Try
            conn.Open()
            Dim cmd As New MySqlCommand("SELECT * FROM `tbl_curd` WHERE `NOPRODUK` LIKE @search OR `NAMAPRODUK` LIKE @search", conn)
            cmd.Parameters.AddWithValue("@search", "%" & txt_search.Text & "%")
            dr = cmd.ExecuteReader
            While dr.Read
                DataGridView1.Rows.Add(dr.Item("NOPRODUK"), dr.Item("NAMAPRODUK"), dr.Item("HARGA"), dr.Item("GRUP"), dr.Item("EXPDATE"), Format(CBool(dr.Item("STATUS"))))
            End While
            dr.Close()
        Catch ex As Exception
            MsgBox("Error loading data: " & ex.Message)
        Finally
            conn.Close()
        End Try
    End Sub

End Class

File: Form1.cs
using System;
using System.Windows.Forms;
using MySql.Data.MySqlClient;

namespace ManajemenStokProduk
{
    public partial class Form1 : Form
    {
        // Inisialisasi objek DatabaseConnection untuk mengelola koneksi ke database
        private DatabaseConnection db = new DatabaseConnection();
        private MySqlConnection conn;

        public Form1()
        {
            InitializeComponent();
        }

        // Event saat Form1 pertama kali dimuat
        private void Form1_Load(object sender, EventArgs e)
        {
            conn = db.GetConnection();

            try
            {
                conn.Open();
                conn.Close();
            }
            catch (MySqlException ex)
            {
                MessageBox.Show("Gagal koneksi : " + ex.Message);
            }

            LoadRecord();
        }

        // Fungsi untuk memuat data dari database ke DataGridView
        public void LoadRecord()
        {
            dataGridView1.Rows.Clear();
            conn = db.GetConnection();  
            conn.Open();
            MySqlCommand cmd = new MySqlCommand("SELECT NOPRODUK, NAMAPRODUK, HARGA, GRUP, EXPDATE, STATUS FROM tbl_crud", conn);
            MySqlDataReader dr = cmd.ExecuteReader();

            // Menghapus semua baris di DataGridView sebelum menambahkan data baru
            while (dr.Read())
            {
                // Mengambil nilai dari DataReader dan menambahkannya ke DataGridView
                bool status = dr["STATUS"].ToString() == "Aktif";
                dataGridView1.Rows.Add(
                    dr["NOPRODUK"].ToString(),
                    dr["NAMAPRODUK"].ToString(),
                    dr["HARGA"].ToString(),
                    dr["GRUP"].ToString(),
                    Convert.ToDateTime(dr["EXPDATE"]).ToString("yyyy-MM-dd"),
                    status
                );
            }

            dr.Close();
            conn.Close();
        }

        // Fungsi untuk mengosongkan inputan
        public void clear()
        {
            // Mengosongkan semua inputan
            txt_noproduk.Clear();
            txt_namaproduk.Clear();
            txt_harga.Clear();
            combo_grupproduk.SelectedIndex = -1;
            dateTimePicker_expdate.Value = DateTime.Today;
            checkbox_status.Checked = false;
        }

        // Event saat tombol Simpan diklik
        private void btn_save_Click(object sender, EventArgs e)
        {
            if (string.IsNullOrWhiteSpace(txt_noproduk.Text) ||
                string.IsNullOrWhiteSpace(txt_namaproduk.Text) ||
                string.IsNullOrWhiteSpace(txt_harga.Text) ||
                string.IsNullOrWhiteSpace(combo_grupproduk.Text) ||
                dateTimePicker_expdate.Value.Date < DateTime.Today)
            {
                MessageBox.Show("Harap lengkapi semua data dengan benar!");
                return;
            }

            try
            {
                // Menyimpan data ke database
                conn = db.GetConnection();
                conn.Open();
                MySqlCommand cmd = new MySqlCommand("INSERT INTO tbl_crud (NOPRODUK, NAMAPRODUK, HARGA, GRUP, EXPDATE, STATUS) VALUES (@NOPRODUK, @NAMAPRODUK, @HARGA, @GRUP, @EXPDATE, @STATUS)", conn);
                
                cmd.Parameters.AddWithValue("@NOPRODUK", txt_noproduk.Text);
                cmd.Parameters.AddWithValue("@NAMAPRODUK", txt_namaproduk.Text);
                cmd.Parameters.AddWithValue("@HARGA", txt_harga.Text);
                cmd.Parameters.AddWithValue("@GRUP", combo_grupproduk.Text);
                cmd.Parameters.AddWithValue("@EXPDATE", dateTimePicker_expdate.Value.ToString("yyyy-MM-dd"));
                cmd.Parameters.AddWithValue("@STATUS", checkbox_status.Checked ? "Aktif" : "Tidak Aktif");

                cmd.ExecuteNonQuery();
                MessageBox.Show("Data berhasil disimpan!");
            }
            catch (Exception ex)
            {
                MessageBox.Show("Gagal menyimpan data: " + ex.Message);
            }
            finally
            {
                conn.Close();
            }
            LoadRecord();   
            clear();
        }

        // Event saat tombol Edit diklik
        private void btn_edit_Click(object sender, EventArgs e)
        {
            if (string.IsNullOrEmpty(txt_noproduk.Text))
            {
                MessageBox.Show("Pilih data yang mau diedit!");
                return;
            }

            try
            {
                // Mengupdate data di database
                conn = db.GetConnection();
                conn.Open();
                MySqlCommand cmd = new MySqlCommand("UPDATE tbl_crud SET NAMAPRODUK = @NAMAPRODUK, HARGA = @HARGA, GRUP = @GRUP, EXPDATE = @EXPDATE, STATUS = @STATUS WHERE NOPRODUK = @NOPRODUK", conn);

                cmd.Parameters.AddWithValue("@NAMAPRODUK", txt_namaproduk.Text);
                cmd.Parameters.AddWithValue("@HARGA", txt_harga.Text);
                cmd.Parameters.AddWithValue("@GRUP", combo_grupproduk.Text);
                cmd.Parameters.AddWithValue("@EXPDATE", dateTimePicker_expdate.Value.ToString("yyyy-MM-dd"));
                cmd.Parameters.AddWithValue("@STATUS", checkbox_status.Checked ? "Aktif" : "Tidak Aktif");
                cmd.Parameters.AddWithValue("@NOPRODUK", txt_noproduk.Text);

                cmd.ExecuteNonQuery();
                MessageBox.Show("Data berhasil diupdate!");
            }
            catch (Exception ex)
            {
                MessageBox.Show("Gagal update data: " + ex.Message);
            }
            finally
            {
                conn.Close();
            }

            // Memuat ulang data setelah edit
            LoadRecord();
            clear();
        }

        // Event saat tombol Hapus diklik
        private void btn_delete_Click(object sender, EventArgs e)
        {
            if (string.IsNullOrEmpty(txt_noproduk.Text))
            {
                MessageBox.Show("Pilih data yang mau dihapus!");
                return;
            }

            // Konfirmasi penghapusan
            DialogResult result = MessageBox.Show("Yakin mau menghapus data ini?", "Konfirmasi Hapus", MessageBoxButtons.YesNo, MessageBoxIcon.Warning);

            // Jika pengguna mengkonfirmasi, lakukan penghapusan
            if (result == DialogResult.Yes)
            {
                try
                {
                    // Menghapus data dari database
                    conn = db.GetConnection();
                    conn.Open();
                    MySqlCommand cmd = new MySqlCommand("DELETE FROM tbl_crud WHERE NOPRODUK = @NOPRODUK", conn);
                    cmd.Parameters.AddWithValue("@NOPRODUK", txt_noproduk.Text);
                    int rowsAffected = cmd.ExecuteNonQuery();

                    // Jika ada baris yang terpengaruh, berarti data berhasil dihapus
                    if (rowsAffected > 0)
                    {
                        MessageBox.Show("Data berhasil dihapus!");
                    }
                    else
                    {
                        MessageBox.Show("Data tidak ditemukan atau gagal dihapus.");
                    }
                }
                catch (Exception ex)
                {
                    MessageBox.Show("Gagal menghapus data: " + ex.Message);
                }
                finally
                {
                    conn.Close();
                }

                // Memuat ulang data setelah hapus
                LoadRecord();
                clear();
            }
        }

        // Event saat sel di DataGridView diklik
        private void dataGridView1_CellClick(object sender, DataGridViewCellEventArgs e)
        {
            // Pastikan sel yang diklik bukan header
            if (e.RowIndex >= 0)
            {
                // Mengambil data dari sel yang diklik dan menampilkannya di TextBox
                DataGridViewRow row = dataGridView1.Rows[e.RowIndex];
                txt_noproduk.Text = row.Cells[0].Value.ToString();
                txt_namaproduk.Text = row.Cells[1].Value.ToString();
                txt_harga.Text = row.Cells[2].Value.ToString();
                combo_grupproduk.Text = row.Cells[3].Value.ToString();
                dateTimePicker_expdate.Value = Convert.ToDateTime(row.Cells[4].Value.ToString());
                checkbox_status.Checked = (row.Cells[5].Value.ToString() == "True");
            }
        }

        // Event saat tombol Clear diklik
        private void btn_clear_Click(object sender, EventArgs e)
        {
            clear();
        }

        // Event saat teks di TextBox pencarian berubah
        private void txt_search_TextChanged(object sender, EventArgs e)
        {
            dataGridView1.Rows.Clear();
            conn = db.GetConnection();
            conn.Open();

            MySqlCommand cmd = new MySqlCommand(
                "SELECT NOPRODUK, NAMAPRODUK, HARGA, GRUP, EXPDATE, STATUS FROM tbl_crud WHERE NOPRODUK LIKE @search OR NAMAPRODUK LIKE @search", conn);
            cmd.Parameters.AddWithValue("@search", "%" + txt_search.Text + "%");

            // Menjalankan query dan membaca hasilnya
            MySqlDataReader dr = cmd.ExecuteReader();
            while (dr.Read())
            {
                // Mengambil nilai dari DataReader dan menambahkannya ke DataGridView
                bool status = dr["STATUS"].ToString() == "Aktif";
                dataGridView1.Rows.Add(
                    dr["NOPRODUK"].ToString(),
                    dr["NAMAPRODUK"].ToString(),
                    dr["HARGA"].ToString(),
                    dr["GRUP"].ToString(),
                    Convert.ToDateTime(dr["EXPDATE"]).ToString("yyyy-MM-dd"),
                    status
                );
            }
            // Menutup DataReader dan koneksi
            dr.Close();
            conn.Close();
        }

        // Event saat tombol Logout diklik
        private void btn_logout_Click(object sender, EventArgs e)
        {
            // Konfirmasi sebelum keluar
            DialogResult result = MessageBox.Show("Yakin mau keluar?", "Konfirmasi Keluar", MessageBoxButtons.YesNo, MessageBoxIcon.Question);

            if (result == DialogResult.Yes)
            {
                Form2 loginForm = new Form2();  
                loginForm.Show();   
                this.Hide();
            }
        }
    }
}
